File: problems/1339_maximum_product_of_splitted_binary_tree/1339_maximum_product_of_splitted_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    final int MOD = 1_000_000_007;
    List<Long> allSums = new ArrayList<>();

    public int maxProduct(TreeNode root) {
        long totalSum = getSum(root);

        long maxP = 0L;
        for (long s : allSums) {
            long p = s * (totalSum - s);
            if (p > maxP) {
                maxP = p;
            }
        }
        return (int) (maxP % MOD);
    }

    private long getSum(TreeNode node) {
        if (node == null)
            return 0L;

        long curSum = getSum(node.left) + getSum(node.right) + node.val;
        allSums.add(curSum);
        return curSum;
    }
}

File: problems/1339_maximum_product_of_splitted_binary_tree/1339_maximum_product_of_splitted_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        all_sums = []

        def get_sum(node):
            if not node:
                return 0
            
            curr_sum = get_sum(node.left) + get_sum(node.right) + node.val
            all_sums.append(curr_sum)
            return curr_sum
        
        total_sum = get_sum(root)

        max_p = 0
        for s in all_sums:
            p = s * (total_sum - s)
            if p > max_p:
                max_p = p
        
        return max_p % (10**9 + 7)